docs/RFCS/table_descriptor_lease.md

```diff
@@ -0,0 +1,249 @@
+acquired and released. A background goroutine will periodically delete
+expired leases. 
+
+Leases will be granted for a duration measured in minutes (we'll
+assume 5m for the rest of this doc, though experimentation may tune
+this number). A node will acquire a lease before using it in an
+operation and may release the lease when the last local operation
+completes that was using the lease.
+
+The leader of the range containing a table descriptor will gossip the
+most recent version of that table descriptor using the gossip key
+`"table-<descID>"` and the value will be the version number. The
+gossiping of the most recent table versions allows nodes to
+asynchronously discover when a new table version is written. But note
+that it is not necessary for correctness as the protocol for acquiring
+a lease ensures that only the most recent version of a descriptor can
+have a new lease granted on it.
+
+Lease acquisition will perform the following steps in a transaction:
+
+* `SELECT descriptor FROM system.descriptor WHERE id = <descID>`
+* `INSERT INTO system.lease VALUES (<desc.ID>, <desc.version>, <nodeID>, <expiration>)`
+
+Nodes will maintain a map from `<descID, version>` to
+`<TableDescriptor, expiration, localRefCount>`. The local reference
+count will be incremented when a transaction first uses a table and
+decremented when the transaction commits/aborts. When the node
+discovers a new version of the table, either via gossip or by
+acquiring a lease and discovering the version has changed it can
+release the lease on the old version when there are no more local
+references:
+
+* `DELETE FROM system.lease WHERE (DescID, Version, NodeID, Expiration) = (<descID>, <version>, <nodeID>, <expiration>)`
+
+A schema change operation needs to ensure that there is only one
+version of a descriptor in use in the cluster before a mutation of the
+descriptor. The operation will perform the following steps
+transactionally:
+
+* `SELECT descriptor FROM system.descriptor WHERE id = <descID>`
+* `SELECT DISTINCT version FROM system.lease WHERE descID = <descID> AND expiration > current_time()`
+* Ensure that there is only one active version which is equal to the current version.
+* Perform the edits to the descriptor and bump the descriptor version.
+* `UPDATE system.descriptor WHERE id = <descID> SET descriptor = <desc>`
+
+Note that the updating of the table descriptor will cause the table
+version to be gossipped alerting nodes to the new version and causing
+them to release leases on the old version. The expectation is that
+nodes will fairly quickly transition to using the new version and
+release all leases to the old version allowing another step in the
+schema change operation to take place. The schema change operation
+must poll the `system.lease` table waiting for the number of active
+versions of a descriptor to fall to 1 or 0. This polling will be
+performed at a short interval (1s), perhaps with a small amount of
+back-off.
```

Comment:
I think the polling might start even lower and have some exponential backoff: 10ms, 20ms, 40ms...5s.

@spencerkimball has promised to think about a redesign for gossip that doesn't have the builtin delay. Gossip as it is currently written is a bit of an odd duck. I'd like to envision it as an efficient RPC fanout mechanism instead. When something is "written to gossip" it would immediately send that RPC to the peers who would forward the message to any peers who hadn't received the message yet. I'm not quite sure how to accomplish this, but its outside of the scope of this RFC.